syntaxes/src/expression.ts

```diff
@@ -0,0 +1,892 @@
+        },
+        {
+          include: '#parenExpression',
+        },
+        {
+          include: '#punctuationComma',
+        },
+        {
+          include: '#punctuationAccessor',
+        },
+      ],
+    },
+
+    arrayLiteral: {
+      name: 'meta.array.literal.ts',
+      begin: /\[/,
+      beginCaptures: {
+        0: {
+          name: 'meta.brace.square.ts',
+        },
+      },
+      end: /\]/,
+      endCaptures: {
+        0: {
+          name: 'meta.brace.square.ts',
+        },
+      },
+      patterns: [
+        {
+          include: '#ngExpression',
+        },
+        {
+          include: '#punctuationComma',
+        },
+      ],
+    },
+
+    booleanLiteral: {
+      patterns: [
+        {
+          name: 'constant.language.boolean.true.ts',
+          match: /(?<!\.|\$)\btrue\b(?!\$)/,
+        },
+        {
+          name: 'constant.language.boolean.false.ts',
+          match: /(?<!\.|\$)\bfalse\b(?!\$)/,
+        },
+      ],
+    },
+
+    expressionOperator: {
+      patterns: [
+        {
+          match: /((?<!\|)(\|)(?!\|))\s?([a-zA-Z0-9\-\_\$]*)/,
+          captures: {
+            2: {
+              name: 'keyword.operator.logical.ts',
+            },
+            3: {
+              name: 'entity.name.function.pipe.ng',
+            },
+          },
+        },
+        {
+          name: 'storage.type.ts',
+          match: /(?<!\.|\$)\b(let)\b(?!\$)/,
+        },
+        {
+          name: 'keyword.control.flow.ts',
+          match: /(?<!\.|\$)\b(await)\b(?!\$)/,
+        },
+        {
+          name: 'keyword.operator.expression.delete.ts',
+          match: /(?<!\.|\$)\bdelete\b(?!\$)/,
+        },
+        {
+          name: 'keyword.operator.expression.in.ts',
+          match: /(?<!\.|\$)\bin\b(?!\$)/,
+        },
+        {
+          name: 'keyword.operator.expression.of.ts',
+          match: /(?<!\.|\$)\bof\b(?!\$)/,
+        },
+        {
+          name: 'keyword.control.if.ts',
+          match: /(?<!\.|\$)\bif\b(?!\$)/,
+        },
+        {
+          name: 'keyword.control.else.ts',
+          match: /(?<!\.|\$)\belse\b(?!\$)/,
+        },
+        {
+          name: 'keyword.control.then.ts',
+          match: /(?<!\.|\$)\bthen\b(?!\$)/,
+        },
+        {
+          name: 'keyword.operator.expression.instanceof.ts',
+          match: /(?<!\.|\$)\binstanceof\b(?!\$)/,
+        },
+        {
+          name: 'keyword.operator.new.ts',
+          match: /(?<!\.|\$)\bnew\b(?!\$)/,
+        },
+        {
+          name: 'keyword.operator.expression.void.ts',
+          match: /(?<!\.|\$)\bvoid\b(?!\$)/,
+        },
+        {
+          begin: /(?<!\.|\$)\bas\b(?!\$)/,
+          beginCaptures: {
+            0: {
+              name: 'keyword.control.as.ts',
+            },
+          },
+          end: /(?=$|"|[;,:})\]])/,
+          patterns: [
+            {
+              include: '#type',
+            },
+          ],
+        },
```

Comment:
Sorry for the nitpick — is there any way we could scope this as a variable instead of a type? The `as` keyword in template expressions isn't a typecast, it's an assignment of the preceding expression to a local variable. It's always bugged me that the JS/TS grammar colors it as a type. :P But it's really not a huge deal, and if need be I can open a separate PR for that change.

Reply:
IMO I think having `index as i` having all of the 3 the same color is something that doesn't make it any easier to read... But if you guys think it fits better, I'm up to that. @ayazhafiz what do you think of this?

Reply:
> IMO I think having `index as i` having all of the 3 the same color is something that doesn't make it any easier to read...

The `as` there would still be a keyword, so it would only be `index` and `i` that get the same color (sorry if that wasn't clear). But FWIW, the more common pattern I see in the wild is `*ngIf="(foo$ | async) as foo"`, and with the `async` there being colored as a function there's still plenty of differentiation. 

Either way, `index as i` is functionally equivalent to `let i = index`, so it strikes me as weirdly inconsistent that the variable name would be colored differently depending on the syntax used to make the assignment.

Reply:
Oh, I understand now. From that perspective, the adequate would be having `let` and `as` having the same color, as they do fundamentally the same thing.

In that case, `as` would be in `storage` instead of `keyword`

Reply:
![image](https://user-images.githubusercontent.com/5878866/73845197-dfca5300-4800-11ea-8c2d-03582bc5dc7f.png)

Reply:
I think this is reasonable. Longer term we should spec out a formal grammar for this.

Reply:
> In that case, `as` would be in `storage` instead of `keyword` alongside with `let`

I think VS Code's theme colors all keywords dark blue with the exception of control keywords which get the purple. It's definitely not a control keyword; storage sounds reasonable to me.

But the thing that actually bugs me here is that the identifier _after_ the `as` is colored as a type (green in VS Code's theme) instead of a variable (light blue). My theme colors types with an italicized orange and local variables with a light pink, so the difference is pretty jarring. :P

But I'm also fine with holding off on this until another PR and maybe taking some time to spec out the grammar in the meantime.

Reply:
I think it will be nearly impossible to solve all open issues in this PR, given its existing size. Though I would have preferred incremental additions of the grammar, which would probably elide these quirks, a lot of time has already been put into this PR already. Given this, I think as long as the grammar testing coverage is thorough and looks reasonable, we can go ahead and merge and apply hotfixes as well as a formal grammar later on. We should stay from additions more than a few hundred lines, let alone nearly 3000, moving forward.